### ISSUE TITLE
Irc connection error handling

### ISSUE THREAD

**Issue by Danatobob** [swiftyspiffy/TwitchLib#81]
Hello. I'm currently investigating an issue i have with my twitch bot that uses this library.
After about 10 hours of work (it varies but usually at least once in 24 hours) it stops responding to chat commands. Restarting the bot fixes it but I'm trying to find a solution that would apply to a working bot.
As things are now, almost no functionality of SmartIrc4net is exposed in TwitchClient. So I've used reflection to get to the IrcConnection field while bot is running. When it stopped working again, I've made screenshots before and after restart.
bot doesn't work (note how it says connected=true):
![ircoffline_props](https://cloud.githubusercontent.com/assets/23310014/20224982/b4867200-a849-11e6-8132-f5d272f23839.png)

bot works
![irconline_props](https://cloud.githubusercontent.com/assets/23310014/20224983/b65ed248-a849-11e6-9d44-13ef8b333131.png)

The only real difference is Lag value which is far beyond allowed ping timeout (16m > 5m). Looking at SmartIrc4net code, I think it fires a ConnectionError event on such cases and sets internal property IsConnectionError to true. Now, I have no way of getting that information out of TwitchLib. It doesnt have an event telling me that something is wrong and it's Reconnect method actually calls ircConnection.Connect which throws "Already connected" exception in this case.
I don't know how important this problem is for you to consider fixing it and I'll probably have a working fix soon enough anyway, but I'd like to post suggestions that address this issue:

- Allow IrcConnection.OnConnectionError event to somehow be consumed by the end user, OR subscribe to this event internally and then reconnect or smth.
- Change TwitchClient.Reconnect to actually use IrcConnection.Reconnect instead of Connect. I don't know if this breaks some use cases but this feels like the proper way of reconnecting - IrcConnection.Reconnect disconnects and then connects.
- Maybe expose IrcConnection to the user as well. I think I'm going to lower ping timeout value for my bot and that would require access to ircConnection.

**Comment by swiftyspiffy** [comment]
This is an excellent Github issue. I've had similar issues in the past, but they seem to have sort of fixed themselves? not sure. Occasionally, I still run in to it though.

Regarding your suggestions, I think the first two are good ideas. I'd like to, if possible, keep IrcConneciton not exposed if possible. I'll implement the first two suggestions now, and perhaps we can talk about possible alternatives for the third one.

**Comment by swiftyspiffy** [comment]
That being said, OnConnecitonError can expose IrcConnection via sender param.

**Comment by XuluniX** [comment]
I encountered this problem before i used TwitchLib on my bot using just a naked socket + NetworkStream and now on TwitchLib. It seems to me like the socket in the background hangs up.
Sometimes it happens after a few hours, sometimes after a day. In some cases, OnDisconnected() is invoked, but sometimes the connections gets stuck without any event called.

A workaround would be to check if the server replies to ping-messages sent to it and reconnect on two lost pings in a row.

**Comment by swiftyspiffy** [comment]
First two implemented here: 6d60d1d98cf539e318b185e2f12ac6fc48e606d2

**Comment by swiftyspiffy** [comment]
@XuluniX that fix could be implemented in the library, but I'd prefer to find the reason for it and correct it, if possible.

**Comment by Danatobob** [comment]
whoa didnt expect fast response. It's 4 AM here, I'll definetly try your fixes but after a few hours of sleep. 
thank you

**Comment by Danatobob** [comment]
UPDATE: I'm still waiting for that connection error to happen on my bot. We have restarted the bot a lot in the past days so it did not have enough time to bug out. 
Meanwhile I've noticed another problem: when internet was down, the bot also received ConnectionError event, but this time "IsConnected" was false, so calling IrcConnection.Reconnect threw an exception:

> Meebey.SmartIrc4net.NotConnectedException: The connection could not be disconnected because there is no active connection
   at Meebey.SmartIrc4net.IrcConnection.Disconnect()
   at Meebey.SmartIrc4net.IrcConnection.Reconnect()

It's like the same issue we solved with TwitchLib.Reconnect when IsConnected was true. I guess there is no way around this:
`if (IsIrcConnected) irc.Reconnect() else irc.Connect()` which must be done either in TwitchLib or in the end user code (which means user would then need access to IrcConnection.IsConnected).

Also after looking thoroughly through SmartIrc4net source code I'm doubting that my initial suggestions  would fix the issue: IrcConnection has internal _OnConnectionError event handler which immediately tries either Disconnect() or Reconnect() (which calls Disconnect() anyway). But then how come my first screenshot showed IsConnected = true ? If Reconnect() did it with successful Connect(), which is weird because how come bot is still broken, then I guess calling Reconnect() via my event handler would not fix anything. So my guess is that IsConnectionError property never became true when the Lag went beyond PingTimeout and _OnConnectionError was never called.

**Comment by swiftyspiffy** [comment]
Thanks for the update.

Currently this is my biggest priority as far as this project goes. I have two instances of the library running hoping that one hits the snag, with any luck.

As far as the reconnect issue, I'll look into that now and implement that fix. From my experience, you get to a state where you're technically connected, because you receive the periodic PING. The problem however is that you don't actually receive any chat/events from Twitch. So, as far as SmartIrc4Net knows, you are indeed connected, despite us not seeing any chat/events.

**Comment by Danatobob** [comment]
I'll apply your Reconnect fix to my bot now, thanks. Less reasons to restart bot manually. I'm still unsure though if calling that inside ConnectionError handler is the right idea considering Reconnect should be already called internally by that time.
> From my experience, you get to a state where you're technically connected, because you receive the periodic PING.

But the Lag thing in my screenshot is doing just that: counting time since last PING was sent. And it stops when PONG is received from server.

```
public TimeSpan Lag {
            get {
                return PingStopwatch.Elapsed;
            }
}
```
and then somewhere in the code there is this
```
                        int last_ping_sent = (int)_Connection.PingStopwatch.Elapsed.TotalSeconds;
                        int last_pong_rcvd = (int)_Connection.NextPingStopwatch.Elapsed.TotalSeconds;
                        // determins if the resoponse time is ok
                        if (last_ping_sent < _Connection._PingTimeout) {
                            ...
                        } else {
                            if (_Connection.IsDisconnecting) {
                                break;
                            }
                            // only flag this as connection error if we are not
                            // cleanly disconnecting
                            _Connection.IsConnectionError = true;
                            break;
                        }
}
```
That is why I think that having Lag > PingTimeout should be a reason to have ConnectionError event fired, which in turn should disconnect. I see that sometimes exceptions are just swallowed in that code, so maybe the code just silently threw an exception before getting to Lag > PingTimeout check. Not sure.

Btw maybe this is related: sometimes _it feels like_ bot stops working not long after I've posted something in chat via browser using same account as bot. Most of the time this doesn't break anything and bot continues working normally, but when it stops it seems to be after my post. For example, bot worked ok entire night, posting automated messages to the channel, then at the morning i've replied to another user in the channel, and like 20 minutes later bot stopped working. I've already tried to reproduce this by posting random stuff while bot is running but like I said most of the time nothing happens and bot continues working, so maybe its just a coincidence.

**Comment by XuluniX** [comment]
I have an instance of a chatbot where the connection is stalled.
Sadly i can't debug is while running on mono.
Using netstat to list the active connections returns the following for the affected process:
`tcp        0      0 188.165.216.204:59049   52.25.8.48:6667         CLOSE_WAIT  0          3480812147  30065/mono
`

However, i have an ZNC instance running on the same machine and i never noticed any problems with the twitch connections. Sometimes there are reconnects but i haven't encountered stalled connections so far.

**Comment by swiftyspiffy** [comment]
I too run my applications using mono which makes it difficult to debug. The instance I had running locally was snagged, but that only happened after I had an unrelated run time error which caused the debugger to pause the application which likely contributed to the snag.

This is a frustrating issue, none the less. The only time I really hit the snag is when library is in production running via mono.

**Comment by swiftyspiffy** [comment]
@XuluniX Can I ask which version of Mono you're using? The CLOSE_WAIT should never happen, and it may be a lead as to what is causing the issue.

**Comment by Danatobob** [comment]
> I had the understanding we should only be receiving PING, and we'd respond with PONG, but maybe I don't understand it.

Hmm, my understanding is the opposite. This is taken from IrcConnection.cs:
```
// determines if it need to send another ping yet
if (last_pong_rcvd > _Connection._PingInterval) {
    _Connection.NextPingStopwatch.Stop();
    _Connection.PingStopwatch.Reset();
    _Connection.PingStopwatch.Start();
    _Connection.WriteLine(Rfc2812.Ping(_Connection.Address), Priority.Critical);
 } 
```
It seems to only send PING and waits for PONG. 
By the way, I'm running bot on windows.
Regarding "hard to debug on mono", If all you need is checking the state of IrcConnection properties, then I'd like to suggest what I'm using: [PropertyGrid](https://msdn.microsoft.com/en-us/library/system.windows.forms.propertygrid(v=vs.110).aspx) . My screenshots above are actually taken from my chatbot UI form, which has PropertyGrid control in "Debug" tab. 
This is a simplified example of what I'm doing:


```

[TypeConverter(typeof(ExpandableObjectConverter))]
public class TwitchLibContainer {

	[TypeConverter(typeof(ExpandableObjectConverter))]
	public IrcConnection IrcConnection { get; private set; }

	private static readonly PropertyInfo piIsIrcConnectionError = typeof(IrcConnection).GetRuntimeProperties().Where(pi => pi.Name == "IsConnectionError").FirstOrDefault();
	public bool IsIrcConnectionError
	{
		get
		{
			return IrcConnection != null && piIsIrcConnectionError != null ? (bool)piIsIrcConnectionError.GetValue(IrcConnection) : false;
		}
	}

	[TypeConverter(typeof(ExpandableObjectConverter))]
	public TwitchClient Client { get; private set; }

	...
	IrcConnection = typeof(TwitchClient).GetFields(BindingFlags.NonPublic | BindingFlags.Instance).Where(f => f.Name == "_client").FirstOrDefault()?.GetValue(Client) as IrcConnection;
	...
}

TwitchLibContainer tlc = new TwitchLibContainer();
debugPropertyGrid.SelectedObject = new { MyTwitchLibContainerProperty = tlc };

...

private void debugRefreshTimer_Tick(object sender, EventArgs e)
{
    debugPropertyGrid.Refresh();
}
```

**Comment by swiftyspiffy** [comment]
So I've enabled AutoRety and set limit to 0 (which is unlimited retries). I have yet to hit the snag described in multiple posts above. I'm not going to push the code for a few days as I want to be sure that the issue was indeed fixed with it. Looks promising though. Both Mono and Windows environments.

**Comment by Danatobob** [comment]
I've also been interested in using AutoRetry. Kinda dislike that its done via recursion (Connect -> Connect) but I guess i would not let bot work without internet for a week or so anyway (my estimate for stack overflow to happen).

**Comment by Danatobob** [comment]
I've updated the bot today with AutoRetry, removed my Reconnect call in OnConnectionError event and just had a lockup again. I forgot to make a screenshot but I've checked that AutoRetry=true and there was AutoRetryAttempt=0 (I assume its a counter). IsConnectionError was true but logs show no irc events (Connected/Disconnected/ConnectionError) in the last 6 hours (and bot was working 30 min ago), so it seems that it only sets the property but doesnt get to call the OnConnectionError event.
That is all I have for now. I guess the issue we are facing is not solved by AutoRetry. I'm wondering if my experimental Reconnect in OnConnectionError event was actually a working fix for this, since I didn't have problems yesterday. Maybe just a coincidence though.

**Comment by Danatobob** [comment]
It seems after I've reconnected the bot from the program, it failed again shortly. 8 hours lag:

![ircoffline1](https://cloud.githubusercontent.com/assets/23310014/20592713/c4ea13d4-b237-11e6-9a01-717dbb13a377.png)
you can see at the bottom IsIrcConnectionError = true. Yet my logs show no irc events since I've reconnected the bot (during reconnect it was Disconnected and Connected).

**Comment by swiftyspiffy** [comment]
If nothing else, we have an indicator of failure (IsIrcConnectionError=true), and I can just implement a timer that checks that flag every few seconds or so. Ideally we wouldn't need a timer though.

**Comment by Danatobob** [comment]
Yea sry, IsAttached is completely unrelated. I had to include it in the screenshot to show IsIrcConnectionError. Im planning to check if AutoConnectError fires or not, and if nothing I'll try dowloading SmartIrc4net sources and enabling logging.

**Comment by swiftyspiffy** [comment]
Sounds good, I might take your code from above and run it on a timer and see if it can detect exactly when the connection hangs up

**Comment by Danatobob** [comment]
The bot bugged again. I think after AutoRetry it happens more frequently. Maybe the code handles errors better when it doesnt go to "if(AutoRetry)" route. I've enabled SmartIrc4net logging now. Next time I should have more info.

**Comment by swiftyspiffy** [comment]
Thanks again for your dedication to this issue. I've been running a chat logging application that is using a build of TwitchLib that has AutoRetry=true and AutoRetryLimit=0 and it has yet to crash. It's a very very simple application. My bot however using the same build does get bugged, but is significantly more complex. Sometimes I think a runtime exception causes a crash within the library or something.

**Comment by Danatobob** [comment]
Finally it bugged and I have logs.
![ircoffline3](https://cloud.githubusercontent.com/assets/23310014/20653202/3161d7a4-b512-11e6-9d5b-c3b70ac19d72.png)
[irc_errorlog.txt](https://github.com/swiftyspiffy/TwitchLib/files/615163/irc_errorlog.txt)
I've replaced my twitch account with mytwitchacc in logs.
Screenshot was taken at about 01:53 AM. 
I think the bug happens at 00:25:39,113. Usually we see this PING-PONG:

```
2016-11-28 00:21:38,899 [IdleWorkerThread (irc.chat.twitch.tv:6667)] DEBUG SOCKET - sent: "PING irc.chat.twitch.tv"
2016-11-28 00:21:39,112 [ReadThread (irc.chat.twitch.tv:6667)] DEBUG SOCKET - received: ":tmi.twitch.tv PONG tmi.twitch.tv :irc.chat.twitch.tv"
2016-11-28 00:21:39,112 [14] DEBUG QUEUE - read: ":tmi.twitch.tv PONG tmi.twitch.tv :irc.chat.twitch.tv"
```

but at 00:25:39 we see this:

```
2016-11-28 00:25:38,901 [IdleWorkerThread (irc.chat.twitch.tv:6667)] DEBUG SOCKET - sent: "PING irc.chat.twitch.tv"
2016-11-28 00:25:39,113 [ReadThread (irc.chat.twitch.tv:6667)] DEBUG SOCKET - received: ":tmi.twitch.tv PONG tmi.twitch.tv :irc.chat.twitch.tv"
2016-11-28 00:28:03,216 [ReadThread (irc.chat.twitch.tv:6667)] DEBUG SOCKET - received: "PING :tmi.twitch.tv"
2016-11-28 00:31:38,903 [IdleWorkerThread (irc.chat.twitch.tv:6667)] WARN SOCKET - ping timeout, connection lost
```
for some reason DEBUG QUEUE doesn't react to PONG.

**Comment by swiftyspiffy** [comment]
Excellent debugging, I have yet to hit a snag, but I know the problem isn't fixed. When I get home, I'll look more into this.

**Comment by Danatobob** [comment]
Yea I'm also thinking about creating an issue at SmartIrc4Net github. I'm not the author of that lib, so I expect that figuring out what exactly went wrong would drain a lot of time and energy ;).

**Comment by swiftyspiffy** [comment]
I hear you on the time and effort. If it's not able to be resolved via SmartIrc4net, I might look into using a different IRC library or writing core IRC functionality myself to remove the dependency. Regardless, I will go create that issue now, and edit this with a link when I am done.

**Comment by Danatobob** [comment]
Nice. Glad that you are willing to fix this one way or another. Let's hope the easy route (SmartIrc4net author looks into issue) will work though.

**Comment by swiftyspiffy** [comment]
I'm going to be testing his suggestion in the referenced issue using the following code, will report back results.

```
Task.Factory.StartNew(() =>
{
    while (true)
    {
        log("Starting listen..");
        _client.Listen();
        log("Stopped listening..");
    }
                    
}).ContinueWith(exStack =>
{
    log("Exception!!");
    exStack.Exception.Flatten().Handle(exception =>
    {
        log(exception.Message);
        return true;
    });
}, TaskContinuationOptions.OnlyOnFaulted);
```

**Comment by Danatobob** [comment]
Allright. I'm currently doing something similar:
```
Task.Factory.StartNew(() => 
            {
                try
                {
                    logger.Debug("Starting listening or irc connection");
                    _client.Listen();
                    logger.Debug("Finished listening or irc connection");
                }
                catch(Exception ex)
                {
                    logger.Fatal("Listen failed", ex);
                }
            });
```
Bot worked 20 hours so far. Waiting for it to bug. And yeah I'm also using github version of SmartIrc4net since that was the only way to enable logging.

**Comment by swiftyspiffy** [comment]
As a sort of update, one bot is at 18 hours, 8 minutes uptime, the other is at 6 hours.

**Comment by swiftyspiffy** [comment]
This code appears to have had a positive effect on the stability of the listen thread. I'm going to push my code (along with some other stuffs). I am also following the comments in the other issue (in SmartIrc4net). Will work from there.

**Comment by Danatobob** [comment]
Allright. Did you figure out why there was an exception in your code ? It was in JoinedChannels.Single() i believe.

**Comment by swiftyspiffy** [comment]
I was not aware of that, but will look into it now.

**Comment by swiftyspiffy** [comment]
Oh wow, I thought Single did something else, oops. I can see how that would throw an exception. 

Fixing now.

**Comment by Danatobob** [comment]
Okay so now we should not receive an exception there. It would still process messages twice if there are 2 Listen methods in parallel, correct ?
What do think about my suggestion in SmartIrc issue about creating a variable for Listen thread and check if it is still running, so that another Connected event does not spawn new thread? Also I'm not sure when your while(true) cycle is supposed to stop when we are disconnecting - should it be due to an exception because Listen would probably not work on a disconnected IrcConnection?
Perhaps you are like me, waiting for SmartIrc author to suggest the correct usage of the library, before writing more code ? Though I'm kinda becoming impatient since my bot crashed a lot in the past days, so maybe I will try your recent changes with some of my ideas regarding single dedicated thread soon.

**Comment by swiftyspiffy** [comment]
Ya, two Listen methods in parallel isn't "correct" (though I haven't had an issue thus far). 
The listen thread is a good idea. Looking back, I actually wasn't thinking about the use case of disconnecting intentionally.
I am sort of waiting, but he seems to be really busy, and am sort of surprised he responded so quick. I'll see if I can implement that Listen thread and some kind of solution to allow for disconnecting.

As a side note, because the library no longer references a nuget package for SmartIrc4net, there were some CI build problems. I fixed it by building the SmartIrc4net dll, and just creating a Libs directory, putting it in it, and creating an exception in the gitingore to allow uploading that dll. Not sure if that is the correct solution, but it worked.

**Comment by Danatobob** [comment]
Btw, I wanted to ask about this for a long time but hesitated to create a separate issue ;).
Why is TwitchLib dependent on log4net 1.2.10 ? It was a painful experience to make my bot work with new log4net and also allow TwitchLib to use a file named log4net_1.2.10.dll, so there would be no conflict. I thought that maybe its for some backward compatibility or smth, but I don't think I saw any usage of log4net in TwitchLib sources.

**Comment by swiftyspiffy** [comment]
Log4net came along with SmartIrc4net. When I first created the project, I didn't know a whole lot about dependencies in .net, so it was kind of "it works, I won't touch it" thing. It should be removed now though. Still learning a lot, just a college student :).

Here is the code I have come up with for the listen thread. `_listenThread` is a global private variable.
```
if(_listenThread == null || _listenThread.IsCompleted)
{
    if (_logging)
        Common.Log("Starting listen..", false, false, Enums.LogType.Success);
    _listenThread = Task.Factory.StartNew(() =>
    {
        try
        {
            _client.Listen();
        } catch(Exception ex)
        {
            if (_logging)
            {
                Common.Log("Exception!!", true, true, Enums.LogType.Failure);
                Common.Log(ex.Message, false, false, Enums.LogType.Failure);
            } 
        }
    });
    if (_logging)
        Common.Log("Stopped listening..", true, true, Enums.LogType.Failure);
} else
{
    if (_logging)
        Common.Log("Already listening, not starting new thread...");
}
```
Obviously `Disconnect` will need to play a role too. What are your thoughts though?

**Comment by Danatobob** [comment]
Yeah this is better in my opinion. As I understood, "Listen" finishes normally only when "IsConnected" (inside SmartIrc) becomes false, so making a cycle is not required since next reconnect will call Connected again anyway. 
Can you make a commit with this ?

**Comment by swiftyspiffy** [comment]
Done: 0ce8ae8c9438d6be5dbf4912bb32db1cfa43be42

**Comment by Danatobob** [comment]
Thank you. I hope this should be it for now for this issue.

**Comment by swiftyspiffy** [comment]
New NuGet package with this fix and many others. Will leave this issue open for a few days to see if it has been fixed.

Cheers.

**Comment by adamcsills** [comment]
<3

**Comment by adamcsills** [comment]
(Have been watching this issue to throw out my home-grown twitch chat solution for yours, happy to see all this activity)

**Comment by swiftyspiffy** [comment]
Ya, this was a difficult one, the most time I've spent on a single issue anyways. Hopefully the library works out for ya :)

**Comment by swiftyspiffy** [comment]
Going to go ahead and close this. Might be a bit trigger happy, but if anyone hits this issue again, feel free to reopen it. Lots of valuable info in this issue.

**Comment by Danatobob** [comment]
I've just stumbled on an issue with the thread variable solution:
I've stopped and started the bot (didn't close the program) and then noticed that it replies twice to my commands. Looking through the logs I've noticed that previous QUEUE thread is still alive and reacting to events. The issue is the following:
When I'm stopping the bot, I do this:
```
if (Client.IsConnected)
  Client.Disconnect();
Client = null;
```
When I start the bot I create a new instance of TwitchClient which obviously doesn't see previous instance's thread when connecting. Another problem is that IrcConnection doesnt seem to stop working when Disconnect is called - logs show that it reconnected back. Maybe AutoRetry is doing that, idk. Also since Listen is blocking, I dont think TwitchLib has any control over when it will stop listening. Though just stopping listening would probably produce spam like in the end of my recent irc logs. So we need a reliable way of stopping irc connection. 
I guess this problem should be easy to reproduce. I'm currently so fed up with irc problems ;)
Other than that, the bot worked absolutely reliably in the past days.

**Comment by swiftyspiffy** [comment]
Interesting, will reopen this. I could have sworn I was receiving confirmation that the `Listen()` had unblocked on `Disconnect()` call.

**Comment by Danatobob** [comment]
I think the problem might be related to the fact that we had internet issues when we decided to reconnect the bot. We switched from one provider to another and then start/stop-ed the bot.  I think internally irc connection might have failed with IsConnectionError=true which causes it to disconnect.
```
2016-12-05 18:33:44,983 [WriteThread (irc.chat.twitch.tv:6667)] WARN SOCKET - connection lost
2016-12-05 18:33:44,986 [WriteThread (irc.chat.twitch.tv:6667)] DEBUG SOCKET - WriteThread aborted
2016-12-05 18:33:44,993 [3] WARN MyTwitchLibDisconnectedEventHandler - Disconnected: mywitchacc, IsIrcConnectionError: True
2016-12-05 18:33:44,995 [3] INFO CONNECTION - disconnected
```
So calling disconnect during that might do nothing because disconnect+IsConnectionError is one of the conditions for AutoRetry i believe.

**Comment by swiftyspiffy** [comment]
We can do that. I'll push that now. Sounds like a promising idea.

**Comment by Danatobob** [comment]
Thank you.
Though after studying SmartIrc sources some more, i think last solution will not always help.
Inside Reconnect() it does Disconnect first, which waits for internal threads to stop working. If we call Disconnect at _that_ time, it will do nothing, because Reconnect will later proceed to Connect again anyway.
I see that smuxi version doesnt use AutoReconnect, but loops Listen and on exception calls Reconnect manually. It also aborts that thread when full stop is required. Either we copy that, or we can try this:
make additional variable inside TwitchClient that specifies if connecting is allowed. Inside Connect() it will be set to true, inside Disconnect() to false. And then if we receive OnConnected event when connection is not allowed, we call Disconnect() immediately. The latter is easier to do but kinda looks lazy, the former would require you to study smuxi version to make it properly though.

**Comment by swiftyspiffy** [comment]
Sorry, I haven't had a ton of time to work on this besides fixing the snafu that happened a day or two ago with NuGet.

Regarding the Disconnect() inside Reconnect(), I am not seeing this, perhaps you can give me a line number. Ideally, I think we should loop Listen() and handle exceptions with a Reconnect() as you mentioned Smuxi does.  I'll see if I can get that done in a few days, I hope. Really busy at the moment.

**Comment by swiftyspiffy** [comment]
As a side note, I just got a random disconnect again. Frustrating.

**Comment by Danatobob** [comment]
I've also switched to other parts of my project and kinda forgot about this issue. Bot works stable so far, even with random internet drops. I don't touch it with stop/start to make sure nothing goes wrong.

> Regarding the Disconnect() inside Reconnect(), I am not seeing this, perhaps you can give me a line number. 

https://github.com/meebey/SmartIrc4net/blob/master/src/IrcConnection/IrcConnection.cs#L794

> As a side note, I just got a random disconnect again.

Are there some new errors in your logs ?

**Comment by swiftyspiffy** [comment]
So I added the while loop back, and it seems to have fixed the problem. I know it's not the correct way to go about it, but it works for the moment.

```
while(true)
{
    try
    {
        _client.Listen();
    }
    catch (Exception ex)
    {
        if (_logging)
        {
            Common.Log("Exception!!", true, true, Enums.LogType.Failure);
            Common.Log(ex.Message, false, false, Enums.LogType.Failure);
        }
    }
}
```

**Comment by Danatobob** [comment]
Allright, but thats not the way to go (like you said) if we are to actually solve this issue. Do you plan on investigating the root cause someday, so that the client doesnt have an unstoppable Listen cycle ?

**Comment by swiftyspiffy** [comment]
Yes. I have not pushed the code above to the repo, yet. I have school finals so I have not been able to work on it as much as I'd like.

The next couple weeks I will be able to investigate it more thoroughly.

**Comment by swiftyspiffy** [comment]
Hey, sorry again about the delay on this. Finally free :)

Pushed this code: 6cd84ea1552505fd34390bb8acceb1d9da072b39
I tested it under a `Disconnect()` call, I disconnected my machine from the net and the thread paused. Reconnected it and after an attempted send/waited period, it reconnected automatically.
Upon `Disconnect()`, `Reconnect()` seems to work as well.

If you could let me know with that new code, that'd be awesome. I'm going to deploy to production for a few days and see how it goes.

**Comment by swiftyspiffy** [comment]
Super long delay on this, but I think the websocketsharp library that this library now uses (removed smartirc4net) handles disconnects with events, and allows users of TwitchLib to fire off a reconnect if needed. As such, I'll close this, feel free to reopen it.
